File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_04/java/Solution.java
package com.hi.dhl.algorithms.offer._04.java;

/**
 * <pre>
 *     author: dhl
 *     desc  :
 * </pre>
 */
class Solution {
    public boolean findNumberIn2DArray(int[][] matrix, int target) {
        if (matrix == null || matrix.length <= 0 || matrix[0].length <= 0) {
            return false;
        }
        int row = 0;
        int column = matrix[0].length - 1;
        int maxRow = matrix.length - 1;
        while (row <= maxRow && column >= 0) {
            if (matrix[row][column] == target) {
                return true;
            } else if (matrix[row][column] > target) {
                column--;
            } else {
                row++;
            }
        }
        return false;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_04/kotlin/Solution.kt
package com.hi.dhl.algorithms.offer._04.kotlin

/**
 * <pre>
 *     author: dhl
 *     desc  :
 * </pre>
 */

class Solution {
    fun findNumberIn2DArray(matrix: Array<IntArray>, target: Int): Boolean {
        if (matrix == null || matrix.size <= 0 || matrix[0].size <= 0) {
            return false;
        }

        var raw = 0
        var column = matrix[0].size - 1;
        val maxRaw = matrix.size - 1;
        while (raw <= maxRaw && column >= 0) {
            when {
                matrix[raw][column] == target -> return true
                matrix[raw][column] > target -> column--
                else -> raw++
            }
        }
        return false
    }
}